docs/coding.md

```diff
@@ -35,3 +35,50 @@ new Rust file types.
 [dup]: https://shadow.github.io/docs/rust/shadow_rs/host/descriptor/struct.Descriptor.html#method.dup
+
+## linux-api vs libc vs nix vs std
+
+We have four-ish ways of accessing platform functionality and definitions. Roughly from
+lowest-level to highest-level:
+
+* Our [`linux-api`](https://github.com/shadow/shadow/tree/main/src/lib/linux-api)
+crate provides fairly low-level bindings over the Linux kernel headers, and a
+few `nix`-style higher-level wrappers.
+* The [`libc`](https://docs.rs/libc/latest/libc/) crate provides fairly low-level bindings
+of libc standard headers.
+* The [`nix`](https://docs.rs/nix/latest/nix/) crate provides a safer and more Rust-idiomatic
+layer on top of the `libc` crate, as well as adapters for underlying `libc` definitions.
+* Rust's [`std`](https://doc.rust-lang.org/std/) crate provides, well, the standard
+way of interacting with the platform, in a portable and Rust-idiomatic way.
+
+When choosing which one to use:
+
+* For code that will be linked into shadow's
+[shim](https://github.com/shadow/shadow/tree/main/src/lib/shim), prefer `linux-api`.
+We want to minimize, and ideally eliminate, usage of `libc` from the shim. `libc` has
+global state that can easily become corrupted when we use it from the shim, which is `LD_PRELOAD`d into managed programs. This is especially becasue
+much of the shim executes in the context of `SIGSYS` signal handlers, meaning we might already
+be in a non-reentrant, non-[async-signal-safe](https://man7.org/linux/man-pages/man7/signal-safety.7.html) libc function higher in the stack. See also <https://github.com/shadow/shadow/milestone/54>.
+
+* For shadow's syscall handler implementations, also prefer `linux-api`. Since
+we are intercepting and implementing at the syscall level, the Linux kernel type
+and constant definitions are the correct ones to use. While types and constants
+in libc are *often* equivalent to kernel types and constants with the same
+names, there are many known cases  where they aren't, and in general there's no
+guarantee even that one that is consistent today will remain consistent
+tomorrow. See also <https://github.com/shadow/shadow/issues/3007>.
+
+* For code that runs in the shadow process, where we are acting as a "normal" program
+that wants to interact with the kernel, generally prefer the highest-level
+interface that provides the necessary functionality. e.g. when creating worker
+threads in Rust, we generally use `std::thread`; there's no reason to use one of the lower
+level crates. Occasionally we need some functionality not provided in `std` though, in which
+case it makes sense to drop down to one of the lower level crates.
+
+* In tests, any of the above can make sense. Theoretically, making direct
+syscalls, e.g. with the `linux-api` crate or `libc`'s `syscall` function, is the
+most direct and precise way of testing shadow's syscall handlers. On the other
```

Comment:
Maybe add that we can use `nix`, but we should only use it for things we aren't intending to test. For example we might want to use `nix::unistd::close` in a socket send-recv test, since we aren't intending to test `SYS_close`.

Reply:
done